### PR TITLE
Implement vgmanager

### PR DESCRIPTION
**Pull request by sp98**
This PR is based on the Rohan's [PR](https://github.com/red-hat-storage/lvm-operator/pull/22)

- Creates VgManager Image
- Adds utility for reading block devices
- Implements controller for vgManager to create vgs.

Testing on SNO cluster:

```
VgManager Image used : quay.io/sp1098/vgmanager:latest 

$ oc version
Client Version: 4.9.0
Server Version: 4.9.0
Kubernetes Version: v1.22.0-rc.0+894a78b

# sapillai @ localhost in ~ [10:52:00]
$ oc get pv
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM          STORAGECLASS              REASON   AGE
pvc-481b5016-6126-4ed1-a42f-5be48f6d132f   10Gi       RWO            Delete           Bound    default/pvc1   topolvm-provisioner-vg1            74s

# sapillai @ localhost in ~ [10:52:03]
$ oc get pods -n lvm-operator-system
NAME                                  READY   STATUS    RESTARTS      AGE
controller-manager-54c78c8df6-mvzkl   2/2     Running   0             42m
topolvm-controller-58c8f86445-brq4x   4/4     Running   0             41m
topolvm-node-md28g                    4/4     Running   8 (40m ago)   41m
vg-manager-bcklj                      1/1     Running   0             41m

# sapillai @ localhost in ~ [10:52:30]
$ oc get pv
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM          STORAGECLASS              REASON   AGE
pvc-481b5016-6126-4ed1-a42f-5be48f6d132f   10Gi       RWO            Delete           Bound    default/pvc1   topolvm-provisioner-vg1            112s

# sapillai @ localhost in ~ [10:52:41]
$ oc get pvc
NAME   STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS              AGE
pvc1   Bound    pvc-481b5016-6126-4ed1-a42f-5be48f6d132f   10Gi       RWO            topolvm-provisioner-vg1   118s

# sapillai @ localhost in ~ [10:52:46]
$ oc get pods
NAME                                READY   STATUS    RESTARTS   AGE
rhocs-bm5labengblrredhatcom-debug   1/1     Running   0          50m
topo-test-6df5cf5457-m8gnz          1/1     Running   0          2m5s

````
TODO (in follow up PRs)
- Revisit controller implementation of the vgManager daemoset.
- Update ListBlockDevice method to read devices using --json flag

### CODE REVIEW

**Review by nbalacha — CHANGES_REQUESTED** [review]
Initial comment: please reorder commits as follows:
1. blockutil
2. vgmanager implementation (includes main.go changes from current build commit)
3. build (only makefile and dockerfile changes)
4. go mod tidy

File: pkg/internal/blockutil.go [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+	// var output bytes.Buffer
+	var blockDevices []BlockDevice
+
+	deviceFSMap, err := GetDeviceFSMap()
+	if err != nil {
+		return []BlockDevice{}, []string{}, fmt.Errorf("failed to list block devices: %w", err)
+	}
+
+	columns := "NAME,ROTA,TYPE,SIZE,MODEL,VENDOR,RO,RM,STATE,KNAME,SERIAL,PARTLABEL"
+	args := []string{"--pairs", "-b", "-o", columns}
```
Comment (sp98) [inline pkg/internal/blockutil.go:57]: can use `lsblk --json` and marshal the output in the BlockDevices struct.  This will take time to test. So I'll update this method in  follow up PR

File: pkg/internal/blockutil.go [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+	Serial     string `json:"serial,omitempty"`
+	PartLabel  string `json:"partLabel,omitempty"`
+}
+
+// ListBlockDevices using the lsblk command
+func ListBlockDevices() ([]BlockDevice, []string, error) {
+	// var output bytes.Buffer
+	var blockDevices []BlockDevice
+
+	deviceFSMap, err := GetDeviceFSMap()
```
Comment (nbalacha) [inline pkg/internal/blockutil.go:51]: lsblk seems to have an FSTYPE output option. Does this need to be done using blkid?
Can be looked into in a separate PR.

File: pkg/internal/blockutil.go (hunk 1 above)
Comment (nbalacha) [reply pkg/internal/blockutil.go:57]: Sounds good.

File: pkg/vgmanager/vgmanager_controller.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	ControllerName = "vg-manager"
+)
+
+var ConfigFilePath = "/mnt/lvmd/lvmd.yaml"
```
Comment (nbalacha) [inline pkg/vgmanager/vgmanager_controller.go:45]: The topolvm-node init container expects "/etc/topolvm/lvmd.yaml"

File: pkg/vgmanager/vgmanager_controller.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+}
+
+func (r *VGReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.Log = log.FromContext(ctx).WithName(ControllerName)
+	r.Log.Info("reconciling", "lvmcluster", req)
+	res, err := r.reconcile(ctx, req)
+	if err != nil {
+		r.Log.Error(err, "reconcile error")
+	}
+	r.Log.V(1).Info("reconcile complete", "result", res)
```
Comment (nbalacha) [inline pkg/vgmanager/vgmanager_controller.go:70]: What is r.Log.V(1)?

File: pkg/vgmanager/vgmanager_controller.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+		remainingValidDevices, matchingDevices, err = filterMatchingDevices(remainingValidDevices, deviceClass)
+		if err != nil {
+			r.Log.Error(err, "could not filterMatchingDevices")
+			continue
+		}
+		if len(matchingDevices) > 0 {
+			// create/update VG and update lvmd config
+			err = r.addMatchingDevicesToVG(matchingDevices, deviceClass.Name)
+			if err != nil {
+				r.Log.Error(err, "could not prepare volume groups")
```
Comment (nbalacha) [inline pkg/vgmanager/vgmanager_controller.go:151]: Include the name of the VG in the message

**Review by nbalacha — CHANGES_REQUESTED** [review]

File: pkg/internal/blockutil.go (hunk 2 above)
Comment (sp98) [reply pkg/internal/blockutil.go:51]: There were some issues around that.  Need to find the BZ around this.

File: pkg/vgmanager/vgmanager_controller.go [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+	if len(delayedDevices) > 0 {
+		requeueAfter = time.Second * 30
+	}
+	return ctrl.Result{RequeueAfter: requeueAfter}, nil
+}
+
+// filterMatchingDevices returns unmatched and matched blockdevices
+func filterMatchingDevices(blockDevices []internal.BlockDevice, lvmCluster lvmv1alpha1.DeviceClass) ([]internal.BlockDevice, []internal.BlockDevice, error) {
+	// currently just match all devices
+	return []internal.BlockDevice{}, blockDevices, nil
```
Comment (nbalacha) [inline pkg/vgmanager/vgmanager_controller.go:182]: Does this skip the installation disk?

File: pkg/vgmanager/lvm.go [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+	}
+	vgLogger := r.Log.WithValues("vgname", vgName)
+	// for vgextend and vgcreate, args takes the form of VGName,dev1,dev2,dev3...
+	var cmd string
+	args := []string{vgName}
+	for _, device := range matchingDevices {
+		args = append(args, fmt.Sprintf("/dev/%s", device.KName))
+	}
+
+	// vgLogger := r.Log.WithValues("vgname", vgName)
```
Comment (nbalacha) [inline pkg/vgmanager/lvm.go:72]: Please remove this.

File: Dockerfile.vgmanager [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+RUN go mod download
+
+# Copy the go source
+COPY cmd/ cmd/
+COPY api/ api/
+COPY pkg/ pkg/
+COPY controllers/ controllers/
+
+# Build
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -o vgmanager cmd/vgmanager/main.go
```
[the remark this answers is not in the document]
Comment (sp98) [reply Dockerfile.vgmanager:19]: how about we do it in a new PR since we have to do it for the operator image as well ?

File: cmd/vgmanager/main.go [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+	flag.StringVar(&probeAddr, "health-probe-bind-address", ":8081", "The address the probe endpoint binds to.")
+	flag.BoolVar(&developmentMode, "development", false, "enable to enable development")
+	opts := zap.Options{}
+	opts.BindFlags(flag.CommandLine)
+	flag.Parse()
+	opts.Development = developmentMode
+
+	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
+
+	setupLog.V(1).Info("development mode on, debug logs enabled")
```
[the remark this answers is not in the document]
Comment (sp98) [reply cmd/vgmanager/main.go]: done.

File: pkg/internal/blockutil.go [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+	// Children   []BlockDevice `json:"children,omitempty"`
+	Rotational string `json:"rota"`
+	ReadOnly   string `json:"ro,omitempty"`
+	Removable  string `json:"rm,omitempty"`
+	PathByID   string `json:"pathByID,omitempty"`
+	Serial     string `json:"serial,omitempty"`
+	PartLabel  string `json:"partLabel,omitempty"`
+}
+
+// IDPathNotFoundError indicates that a symlink to the device was not found in /dev/disk/by-id/
```
[the remark this answers is not in the document]
Comment (sp98) [reply pkg/internal/blockutil.go]: removed

File: pkg/vgmanager/vgmanager_controller.go (hunk 5 above)
Comment (sp98) [reply pkg/vgmanager/vgmanager_controller.go]: done.

File: pkg/vgmanager/vgmanager_controller.go (hunk 6 above)
Comment (sp98) [reply pkg/vgmanager/vgmanager_controller.go:181]: No. Right now there is no filter. we just return all the devices back.

File: pkg/vgmanager/lvm.go (hunk 7 above)
Comment (sp98) [reply pkg/vgmanager/lvm.go]: done.

File: pkg/vgmanager/vgmanager_controller.go (hunk 4 above)
Comment (sp98) [reply pkg/vgmanager/vgmanager_controller.go]: Not sure. Looks like different verbosity levels. I'll remove it.

File: controllers/vgmanager_daemonset.go [hunk 11]
```diff
@@ -33,12 +33,12 @@ var (
 	DevDirVolName = "device-dir"
 	SysVolName    = "sys"
 
-	LVMdDir    = "/mnt/lvmd/"
+	LVMdDir    = "/etc/topolvm"
```
Comment (jmolmo) [inline controllers/vgmanager_daemonset.go:36]: if this is the directory where the lvmd config file is stored, then this can cause problems, because we already have another constant with the complete file path. So in case of change, everybody must know that we need to change this other constant.
I suggest to use only the "LvmdConfigFile" constant , and if it is needed extract the base path from the constant.

File: pkg/internal/blockutil.go [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+	Name   string `json:"name"`
+	KName  string `json:"kname"`
+	Type   string `json:"type"`
+	Model  string `json:"model,omitempty"`
+	Vendor string `json:"vendor,omitempty"`
+	State  string `json:"state,omitempty"`
+	FSType string `json:"fsType"`
+	Size   string `json:"size"`
+	// Children   []BlockDevice `json:"children,omitempty"`
+	Rotational string `json:"rota"`
```
Comment (jmolmo) [inline pkg/internal/blockutil.go:38]: Having a different names (struct/json) only can drive to confusion. I would suggest to use the same names in both parts... so instead of "rota" use "rotational" , ro/readonly, rm/removable

File: pkg/vgmanager/vgmanager_controller.go (hunk 6 above)
Comment (sp98) [reply pkg/vgmanager/vgmanager_controller.go:181]: I spoke too soon. Yes there is a check.  There is a filter called  `noBiosBootInPartLabel`. It checks for `bios` or `boot` in the disk part label. Based on our experience with LSO so far, it might not cover all the cases.

File: controllers/vgmanager_daemonset.go (hunk 11 above)
Comment (sp98) [reply controllers/vgmanager_daemonset.go:36]: Good point. I would like to take this up in a follow up PR.

File: pkg/internal/blockutil.go (hunk 12 above)
Comment (sp98) [reply pkg/internal/blockutil.go:38]: I would take this is up in a follow up PR while working on [this](https://github.com/red-hat-storage/lvm-operator/pull/27#discussion_r773119630) comment

File: pkg/vgmanager/vgmanager_controller.go [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+		return ctrl.Result{}, err
+	} else if len(badRows) > 0 {
+		r.Log.Error(err, "could not parse all the lsblk rows", "lsblk.BadRows", badRows)
+	}
+
+	existingLvmdConfig := &lvmdCMD.Config{}
+
+	// load lvmd config
+	lvmdConfig := &lvmdCMD.Config{
+		SocketName: topolvm.DefaultLVMdSocket,
```
Comment (leelavg) [inline pkg/vgmanager/vgmanager_controller.go:95]: - Is the value of `topolvm.DefaultLVMdSocket` and what's supplied at https://github.com/red-hat-storage/lvm-operator/blob/7b5eb349b5fd22374b24b6b571238999c6c68e7b/controllers/topolvm_node.go#L277 to nodeplugin same?

Comment (sp98) [reply pkg/vgmanager/vgmanager_controller.go:95]: yes. Same. https://github.com/topolvm/topolvm/blob/af1ff7bb2620d492e50047a041997fd86d55ac48/constants.go#L36

Comment (sp98) [reply pkg/vgmanager/vgmanager_controller.go:95]: not same.  https://github.com/topolvm/topolvm/blob/af1ff7bb2620d492e50047a041997fd86d55ac48/constants.go#L36

Comment (nbalacha) [inline pkg/vgmanager/vgmanager_controller.go:95]: Please replace this with controllers.LVMdSocketPath

**Review by nbalacha — APPROVED** [review]
